### DISCUSSION TITLE
Octoherd script composition

### DISCUSSION THREAD

**Discussion by oscard0m** [octoherd/octoherd#21] [Ideas]
I was asking myself. Would it be possible to compose a set of Octoherd scripts?

A use case I found myself is to apply some setting updates from an existing repository to the rest of repositories. 

For example, for the same query of repositories and using the same repository as template I would like to:
1. Update its branch protection (using https://github.com/octoherd/script-sync-branch-protections) 
2. Update its merge options (using https://github.com/oscard0m/octoherd-script-sync-repo-settings)

Now it can be done running those scripts separately but it would be nice to re-use the same query and apply both changes. Not sure if there would be a possible tweak or way to open the door to this or it is expected to do all this logic from user end.

**Comment by gr2m** [comment]
Yes I created scripts with composition in mind. You should be able to create a script that imports any number of other scripts and pass through the 3 arguments (`octokit`, `repository`, `options`). I never tried it though.

We could also support `octoherd run` to run multiple scripts at once? But that might introduce complexities that we are not aware of, I'd like to implement the other open items first, such as schemas for options, required auth (scopes and permissions) and support for GitHub App auth

**Reply by oscard0m** [reply]
Yep I was considering this second option of implementing scripts which are published and useful by their own and, at the same time, to be able to import them and run them:

Like having these 3 octoherd scripts which can be run independendently:
* `octoherd-script-sync-repository-branch-protection`
* `octoherd-script-sync-repository-teams`
* `octoherd-script-sync-repository-options`

And at the same time, create an octoherd script `octoherd-script-sync-repository-settings` (which internally is calling these 3 scripts, but running the query for getting the list of repos only once.).

---

It's fine on focusing on other stuff first, I just wanted to assure the option did not exist yet.

Thanks

**Reply by gr2m** [reply]
you can go with the first option for now and see if it works, I'd love to see it
